fdbserver/storageserver.actor.cpp

```diff
@@ -3521,56 +3521,85 @@ bool rangeQuery(const std::string& s) {
+				vt.push_back(Optional<Tuple>(escapedTuple));
 			} else if (singleKeyOrValue(s, sz)) {
-				int idx;
-				Tuple* referenceTuple;
-				try {
-					idx = std::stoi(s.substr(3, sz - 5));
-				} catch (std::exception& e) {
-					throw mapper_bad_index();
-				}
-				if (s[1] == 'K') {
-					unpackKeyTuple(&referenceTuple, keyTuple, keyValue);
-				} else if (s[1] == 'V') {
-					unpackValueTuple(&referenceTuple, valueTuple, keyValue);
-				} else {
-					ASSERT(false);
-					throw internal_error();
-				}
-				if (idx < 0 || idx >= referenceTuple->size()) {
-					throw mapper_bad_index();
-				}
-				mappedKeyTuple.append(referenceTuple->subTuple(idx, idx + 1));
+				// when it is SingleKeyOrValue, insert an empty Tuple to vector as placeholder
+				vt.push_back(Optional<Tuple>(Tuple()));
 			} else if (rangeQuery(s)) {
-				if (i != mappedKeyFormatTuple.size() - 1) {
-					// It must be the last element of the mapper tuple
-					throw mapper_bad_range_decriptor();
-				}
-				// Every record will try to set it. It's ugly, but not wrong.
+				// when it is rangeQuery, insert Optional.empty as placeholder
+				vt.push_back(Optional<Tuple>());
 				isRangeQuery = true;
-				// Do not add it to the mapped key.
 			} else {
-				// If the element is a string but neither escaped nor descriptors, add to result.
-				mappedKeyTuple.append(mappedKeyFormatTuple.subTuple(i, i + 1));
+				vt.push_back(Optional<Tuple>(mappedKeyFormatTuple.subTuple(i, i + 1)));
 			}
 		} else {
-			// If the element not a string, add to result.
-			mappedKeyTuple.append(mappedKeyFormatTuple.subTuple(i, i + 1));
+			vt.push_back(Optional<Tuple>(mappedKeyFormatTuple.subTuple(i, i + 1)));
+		}
+	}
+}
+
+Key constructMappedKey(KeyValueRef* keyValue,
+                       std::vector<Optional<Tuple>>& vec,
+                       Tuple& mappedKeyTuple,
+                       std::vector<std::string>& strings) {
+	// Lazily parse key and/or value to tuple because they may not need to be a tuple if not used.
+	Optional<Tuple> keyTuple;
+	Optional<Tuple> valueTuple;
+	mappedKeyTuple.clear();
+	mappedKeyTuple.reserve(vec.size());
+
+	for (int i = 0; i < vec.size(); i++) {
+		if (!vec[i].present()) {
+			// rangeQuery
+			continue;
+		}
+		if (vec[i].get().size()) {
+			mappedKeyTuple.append(vec[i].get());
+		} else {
+			// singleKeyOrValue is true
+			std::string s = strings[i];
+			auto sz = s.size();
+			int idx;
+			Tuple* referenceTuple;
+			try {
+				idx = std::stoi(s.substr(3, sz - 5));
+			} catch (std::exception& e) {
+				throw mapper_bad_index();
+			}
+			if (s[1] == 'K') {
+				unpackKeyTuple(&referenceTuple, keyTuple, keyValue);
+			} else if (s[1] == 'V') {
+				unpackValueTuple(&referenceTuple, valueTuple, keyValue);
+			} else {
+				ASSERT(false);
+				throw internal_error();
+			}
+			if (idx < 0 || idx >= referenceTuple->size()) {
+				throw mapper_bad_index();
+			}
+			mappedKeyTuple.append(referenceTuple->subTuple(idx, idx + 1));
```

Comment:
`subTuple` call is expensive. And since here we only extract one field, we could potentially just extract the underlying serialized `StringRef` and append it to `mappedKeyTuple`.

Reply:
Sounds good.